presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.google.common.collect.ImmutableList;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.concurrent.CompletableFuture.allOf;
+
+public class PartitionedConsumption<T extends Closeable>
+{
+    private final int consumers;
+    private final Function<Integer, CompletableFuture<T>> loader;
+    private final Partition<T> firstPartition;
+
+    PartitionedConsumption(int consumers, Collection<Integer> partitionNumbers, Function<Integer, CompletableFuture<T>> loader)
+    {
+        this.consumers = consumers;
+        this.loader = loader;
+        this.firstPartition = createPartitions(partitionNumbers);
+    }
+
+    private Partition<T> createPartitions(Iterable<Integer> partitionNumbers)
+    {
+        List<Integer> partitionsReversed = ImmutableList.copyOf(partitionNumbers).reverse();
+        checkArgument(!partitionsReversed.isEmpty(), "partitionNumbers can't be empty");
+        Optional<Partition<T>> partition = Optional.empty();
+        for (Integer partitionNumber : partitionsReversed) {
+            partition = Optional.of(new Partition<>(consumers, partitionNumber, loader, partition));
+        }
+        Partition<T> firstPartition = partition.get();
+        firstPartition.previousReleased();
+        return firstPartition;
+    }
+
+    public int getConsumersCount()
+    {
+        return consumers;
+    }
+
+    public Partition<T> firstPartition()
+    {
+        return firstPartition;
+    }
+
+    public static class Partition<S extends Closeable>
```

Comment:
I would split this class to two.

PartitionChain (as suggested by @findepi) with methods:
 * CompletableFuture<Optional<Partition>> loadNextPartition()

And Partition with method:
 * release()